src/backend/distributed/catalog/dependency.c

```diff
@@ -0,0 +1,381 @@
+	ScanKeyInit(&key[1], Anum_pg_depend_objid, BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(target->objectId));
+	depScan = systable_beginscan(depRel, DependDependerIndexId, true, NULL, 2, key);
+
+	while (HeapTupleIsValid(depTup = systable_getnext(depScan)))
+	{
+		Form_pg_depend pg_depend = (Form_pg_depend) GETSTRUCT(depTup);
+		ObjectAddress address = { 0 };
+		ObjectAddressSet(address, pg_depend->refclassid, pg_depend->refobjid);
+
+		if (follow == NULL || !follow(context, pg_depend))
+		{
+			/* skip all pg_depend entries the user didn't want to follow */
+			continue;
+		}
+
+		/*
+		 * recurse depth first, this makes sure we call apply for the deepest dependency
+		 * first.
+		 */
+		recurse_pg_depend(&address, follow, apply, context);
+
+		/* now apply changes for current entry */
+		if (apply != NULL)
+		{
+			apply(context, pg_depend);
+		}
+	}
+
+	systable_endscan(depScan);
+	relation_close(depRel, AccessShareLock);
+}
+
+
+/*
+ * follow_get_dependencies_for_object applies filters on pg_depend entries to follow all
+ * objects which should be distributed before the root object can safely be created.
+ *
+ * Objects are only followed if all of the following checks hold true:
+ *  - the pg_depend entry is a normal dependency, all other types are created and
+ *    maintained by postgres
+ *  - the object is not already in the targetList (context), if it is already in there an
+ *    other object already caused the creation of this object
+ *  - the object is not already marked as distributed in pg_dist_object. Objects in
+ *    pg_dist_object are already available on all workers.
+ *  - object is not created by an other extension. Objects created by extensions are
+ *    assumed to be created on the worker when the extension is created there.
+ *  - citus supports object distribution. If citus does not support the distribution of
+ *    the object we will not try and distribute it.
+ */
+static bool
+follow_get_dependencies_for_object(void *context, const Form_pg_depend pg_depend)
+{
+	List **targetList = (List **) context;
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, pg_depend->refclassid, pg_depend->refobjid);
+
+	if (pg_depend->deptype != DEPENDENCY_NORMAL)
+	{
+		return false;
+	}
+
+	if (IsObjectAddressInList(&address, *targetList))
+	{
+		return false;
+	}
+
+	if (isObjectDistributed(&address))
+	{
+		return false;
+	}
+
+	if (IsObjectAddressOwnedByExtension(&address))
+	{
+		return false;
+	}
+
+	if (!SupportedDependencyByCitus(&address))
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * follow_order_object_address applies filters on pg_depend entries to follow the
+ * dependency tree of objects in depth first order. The filters are practically the same
+ * to follow_get_dependencies_for_object, except it will follow objects that have already
+ * been distributed. This is because its primary use is to order the objects from
+ * pg_dist_object in dependency order.
```

Comment:
I feel like the naming doesn't quite capture the subtle distinction that we're trying to make here.

maybe FollowAllSupportedDependencies vs. FollowNewSupportedDependencies?